dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java

```diff
@@ -171,36 +180,103 @@ private static String[] getArgNameArray(AnnotatedMethod<?> method, Map<String, A
+                    case NONCONCRETE_GENERIC -> genericParams.add(param.getJavaParameter().getName());
                 }
-                String argName = resolveArgumentName(param, argAnnotation);
-                boolean isDuplicateArg = result.containsKey(argName);
-
-                result.put(argName, new ArgumentMetadata(param.getBaseType(),
-                                                         param.getPosition(),
-                                                         argAnnotation.description(),
-                                                         argAnnotation.required(),
-                                                         isDuplicateArg));
-            }
 
+            } else {
+                addArgumentMetadata(param, result);
+            }
         }
         if (!genericParams.isEmpty()) {
             throw new GenericArgumentException(genericParams);
         }
         return result.isEmpty() ? Collections.emptyMap() : result;
     }
 
+    public static Map<String, ArgumentMetadata> getArgumentMap(AnnotatedMethod<?> method) {
+        Map<String, ArgumentMetadata> result = new HashMap<>();
+        for (AnnotatedParameter<?> param : method.getParameters()) {
+            addArgumentMetadata(param, result);
+        }
+        return result.isEmpty() ? Collections.emptyMap() : result;
+    }
+
+    private static void addArgumentMetadata(AnnotatedParameter<?> param, Map<String, ArgumentMetadata> result) {
+        ToolArg argAnnotation = param.getAnnotation(ToolArg.class);
+
+        if (argAnnotation != null) {
+            String argName = resolveArgumentName(param, argAnnotation);
+            boolean isDuplicateArg = result.containsKey(argName);
+
+            result.put(argName, new ArgumentMetadata(param.getBaseType(),
+                                                     param.getPosition(),
+                                                     argAnnotation.description(),
+                                                     argAnnotation.required(),
+                                                     isDuplicateArg));
+        }
+    }
+
+    private static void addArgumentMetadata(AnnotatedParameter<?> param, Map<String, ArgumentMetadata> result, Map<TypeVariable<?>, Type> genericMap) {
+        ToolArg argAnnotation = param.getAnnotation(ToolArg.class);
+
+        if (argAnnotation != null) {
+            String argName = resolveArgumentName(param, argAnnotation);
+            boolean isDuplicateArg = result.containsKey(argName);
+
+            result.put(argName, new ArgumentMetadata(TypeUtility.createCustomType(param.getBaseType(), genericMap),
+                                                     param.getPosition(),
+                                                     argAnnotation.description(),
+                                                     argAnnotation.required(),
+                                                     isDuplicateArg));
+        }
+
+    }
```

Comment:
Could we consolidate these methods so that one calls the other with an empty `genericMap` to reduce duplication?

Additionally, if we always create a custom type when necessary, do we need to pass through the generic map to the schema generator at all?

Reply:
Yeah, maybe that would make sense.

Initially I thought we should keep the tool-specific stuff out of the schema generator, but given that we already have a method which is specific to tool inputs, I think it's ok.